### PR TITLE
[BALDE-19] Input Manager

### PR DESCRIPTION
**Pull request by BIGbadEL**

### CODE REVIEW

**Comment by piotrmoszkowicz** [comment]
There's something wrong with your master repository. Let's talk about it today 😉

File: src/input/input_manager.h [hunk 1]
```diff
@@ -0,0 +1,135 @@
+//
+// Created by grzegorz on 17.12.18.
+//
+
+#pragma once
+
+#include "GLFW/glfw3.h"
+
+#define MAX_KEYS 1024
```
Comment (Blinkuu) [inline src/input/input_manager.h]: Do we really have 1024 keys? Maybe there is some possibility to make it somewhat generic?

File: src/input/input_manager.h [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+#define MAX_MOUSE_BUTTONS 32
```
Comment (Blinkuu) [inline src/input/input_manager.h]: Same, as stated above.

File: examples/input_example.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    GLFWwindow *window = glfwCreateWindow(800, 600, "LearnOpenGL", nullptr, nullptr);
+    glfwMakeContextCurrent(window);
+
+    glfwSetKeyCallback(window, key_callback);
+    glfwSetMouseButtonCallback(window, mouse_button_callback);
+    glfwSetCursorPosCallback(window, cursor_position_callback);
+
+    while (!glfwWindowShouldClose(window)) {
+
+        double x,y;
```
Comment (4jeR) [inline examples/input_example.cpp:49]: Declaring variables in a loop, is it possible to declare them once out of the loop ?

Comment (BIGbadEL) [reply examples/input_example.cpp:49]: doesn't matter with trivial types

**Review by Blinkuu — APPROVED** [review]
good lil' c++ master

File: src/input/input_manager.h [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+         * @fn    Update
+         * @brief function updating information about input events should be called every frame
+         */
+
+        static void Update() noexcept; //TODO: have to be called in Window class Update
+
+        /**
+         * @fn            isKeyPressed
+         * @brief         function that checks if key was pressed
+         * @param keycode GLFW macro e.g. GLfW_KEY_SPACE
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type

File: src/input/input_manager.h [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+         * @return        true if key was pressed otherwise false
+         */
+
+        [[nodiscard]] inline static bool isKeyPressed(int keycode) noexcept;
+
+        /**
+         * @fn            isKeyTyped
+         * @brief         function that checks if key was typed
+         * @param keycode GLFW macro e.g. GLfW_KEY_SPACE
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type

File: src/input/input_manager.h [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+         * @return        true if key was typed otherwise false
+         */
+
+        [[nodiscard]] inline static bool isKeyTyped(int keycode) noexcept;
+
+        /**
+         * @fn               isMouseButtonPressed
+         * @brief            function that checks if mouse button was pressed
+         * @param buttoncode GLFW macro e.g. GLFW_MOUSE_BUTTON_1
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type

File: src/input/input_manager.h (hunk 5 above)
Comment (piotrmoszkowicz) [reply src/input/input_manager.h]: Return type

File: src/input/input_manager.h (hunk 6 above)
Comment (piotrmoszkowicz) [reply src/input/input_manager.h]: Return type

File: src/input/input_manager.h [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+         * @return           true if button was pressed otherwise false
+         */
+
+        [[nodiscard]] inline static bool isMouseButtonPressed(int buttoncode) noexcept;
+
+        /**
+         * @fn               isMouseButtonTyped
+         * @brief            function that checks if mouse button was typed
+         * @param buttoncode GLFW macro e.g. GLFW_MOUSE_BUTTON_1
+         * @return           true if button was typed otherwise false
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type, return type

File: src/input/input_manager.h [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+         * @param xpos [double] x-coordinate
+         * @param ypos [double] y-coordinate
+         */
+
+        inline static void getMousePos(double &xpos, double &ypos) noexcept;
+
+        /**
+         * @fn                    SetKeyPressedCallback
+         * @brief                 sets function which will be called on key pressed event
+         * @param keycode         GLFW key id
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type

File: src/input/input_manager.h [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+         * @param fun [void(*)()] void function with no parameter
+         */
+
+        static void SetKeyPressedCallback(int keycode, callback fun) noexcept;
+
+        /**
+         * @fn                    SetKeyTypedCallback
+         * @brief                 sets function which will be called on key typed event
+         * @param keycode         GLFW key id
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type

File: src/input/input_manager.h [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+         * @param fun [void(*)()] void function with no parameter
+         */
+
+        static void SetKeyTypedCallback(int keycode, callback fun) noexcept;
+
+        /**
+         * @fn                    SetMouseButtonPressedCallback
+         * @brief                 sets function which will be called on mouse button pressed event
+         * @param buttoncode      GLFW mouse button id
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type

File: src/input/input_manager.h [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+         * @param fun [void(*)()] void function with no parameter
+         */
+
+        static void SetMouseButtonPressedCallback(int buttoncode, callback fun) noexcept;
+
+        /**
+         * @fn                    SetMouseButtonTypedCallback
+         * @brief                 sets function which will be called on mouse button Typed event
+         * @param buttoncode      GLFW mouse button id
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Param type

File: src/input/input_manager.h [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+         * @param fun [void(*)()] void function with no parameter
```
Comment (piotrmoszkowicz) [inline src/input/input_manager.h]: Don't we rename it to the callback as it is set?

File: tests/input_test.cpp [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    using namespace Bald;
+    using namespace Input;
+
+
+    glfwInit();
+
+    GLFWwindow* window = glfwCreateWindow(800, 600, "LearnOpenGL", NULL, NULL);
+    glfwMakeContextCurrent(window);
+
+    EXPECT_TRUE(glfwSetKeyCallback(window, key_callback) == NULL);
```
Comment (piotrmoszkowicz) [inline tests/input_test.cpp]: Why not EXPECT_THAT(glfwSet..., NULL) ?
https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#matchers

**Review by piotrmoszkowicz — CHANGES_REQUESTED** [review]
Overall nice, I have some question to the overall idea, but let's talk about it today irl.

**Review by piotrmoszkowicz — APPROVED** [review]

**Pull request merged**